Hopper

Thisisnotgoingtowork opened the pull request:
Some lambdas functionality is added in the Circle, and template functional in DataTracker, and FunctionSet. There are some issue with the Fixpoint, but for the advance class, the test should work no problem. We will be continuing to add things to them.

AmeryJohnson reviewed (commented):
### Code Design
**Informative Identifiers:** Names are clear and understandable.
**Magic Numbers:** In class Circle for functions regenEnergy and eatPreyCircle I am not sure what the numbers 100 and 5 are doing but other than that I believe everything looks good.
**Intuitive and Functional Interface:** I believe everything looks good here the functions seem to be understandable, member variables are private, easily readable.
**Informative Comments:** I think there could definitely be some more comments made, possibly explaining some of the functions especially in DataTracker such as the lambdas being used and STL algorithms as well. I think more comments could be made in general even just some comments about what the private member variables are used for and what they're doing.
**Consistent Style:** I did not notice any inconsistencies with the code style.
**Visual Appearance:** Overall I think the appearance looks good; Could maybe use a bit more whitespace or comments that explain some of the more complicated lines of code.

Overall I think the code design was done pretty well with a few areas that could possibly be improved.

### Code Optimization and Flexibility
Made comments in line for this portion.

**Side Note:** I believe you need to include string in Circle.h because as of right now there are errors relating to using strings. I did end up including string locally to see if the tests passed and they do all pass but there are a couple warnings present.

ephraimbennett reviewed (commented):
## Testing & Error Checking
### Unit testing. 
Unit tests for the DataTracker are well organized and seem to trigger most cases of private portions as well. There could be some other edge cases tested however.
For the advanced Circle class, there are two key unit tests that are commented out. Obviously you'd want to get these up and working.
### Edge cases.
There could be some extra edge cases that are specified below, such as empty containers, adding and removing and THEN calculating, etc.
### Assertions.
Assertions are properly used throughout the test cases
### Error Handling.
There are certain conditions like divide by zero which trigger exceptions that are not verified in the test cases. Specified Below.

ephraimbennett commented on tests/Group-03/DataTrackerAdvanced.cpp:
```diff
@@ -28,18 +29,11 @@ TEST_CASE("Advanced DataTracker Variance", "[DataTracker]") {
     CHECK(tracker.variance() == Approx(4.0));
 }
 
-TEST_CASE("Advanced DataTracker Template Support", "[DataTracker]") {
```
Seems like the templated tests should remain? A goal of the advanced DataTracker as states was to support non-numerical values.

ephraimbennett replied:
Not sure why this was removed, one of the specifications for the advanced DataTracker was to handle non-numerical values.

ephraimbennett commented on tests/Group-03/DataTrackerAdvanced.cpp:
```diff
@@ -28,18 +29,11 @@ TEST_CASE("Advanced DataTracker Variance", "[DataTracker]") {
-}
+
 
 TEST_CASE("Advanced DataTracker Prey-Predator Win Condition", "[DataTracker]") {
```
For another edge case you can try adding and removing values and then recheck win conditions, ie. remove some preys then assert winner() == 0, etc. Also check what happens when the list of data is empty and winner() is called

ephraimbennett replied:
For another edge case you can try adding and removing values and then recheck win conditions, ie. remove some preys then assert winner() == 0, etc. Also check what happens when the list of data is empty and winner() is called

ephraimbennett commented on tests/Group-03/DataTrackerAdvanced.cpp:
```diff
@@ -2,6 +2,7 @@
 #include <vector>
+#define CATCH_CONFIG_MAIN
 
 TEST_CASE("Advanced DataTracker Mode", "[DataTracker]") {
```
For this test case you could also add an instance with negative values. Also like most test cases, add an empty DataTracker test

ephraimbennett replied:
For this test case you could also add an instance with negative values. Also like most test cases, add an empty DataTracker test. It should verify the exceptions thrown from something like divide by zero if that occurs when the data tracker is empty. This applies to the variance one as well.

ephraimbennett commented on tests/Group-03/CircleAdvanced.cpp:
```diff
@@ -43,27 +45,29 @@ TEST_CASE("CircleTest_EnergyDepletionAndRegeneration", "[Circle]") {
-    predator.consumePrey(prey1);
-    CHECK(predator.getConsumedPreyCount() == 1);
-    CHECK(prey1.isConsumed());
+// TEST_CASE("CircleTest_PredatorConsumesPreyAndPopulates", "[Circle]") {
```
Also wanna get this test case working, it's a key part of the advanced class.

ephraimbennett reviewed (commented):
## Code Optimization and Flexibility.
For DataTracker I think there are some marginal optimizations you can make that are specified below.
Circle doesn't seem to be very computationally expensive, but I added my thoughts where possible.

ephraimbennett commented on Group-03/src/DataTracker.cpp:
```diff
@@ -5,104 +5,123 @@
 
     size_t size = sorted_values.size();
-    
     if (size % 2 == 0) {
```
Should wrap this bit with
if constexpr (std::is_arithmetic_v<T>) 
Before static casting into a double, to avoid compilation errors.

Devon-FoxElster commented on Group-03/src/DataTracker.cpp:
```diff
@@ -5,104 +5,123 @@
+// Removes a value if found, returns true if successful
+template <typename T>
+bool DataTracker<T>::delete_value(const T& value) {
     auto it = std::find(values.begin(), values.end(), value);
```
While std::find works well enough in this context I would recommend using the std::ranges version, as it has a simpler function call and more modern implementation

ephraimbennett commented on Group-03/src/DataTracker.cpp:
```diff
@@ -5,104 +5,123 @@
+        double mean_value = mean();
+        double variance_sum = std::accumulate(values.begin(), values.end(), 0.0,
+            [mean_value](double acc, const T& val) {
+                return acc + std::pow(val - mean_value, 2);
```
In this case, you can optimize slightly by using (val - mean_value) * (val - mean_value) instead of std::pow(..., 2). It should be faster with most optimization flags and compilers.

https://stackoverflow.com/questions/2940367/what-is-more-efficient-using-pow-to-square-or-just-multiply-it-with-itself

Devon-FoxElster commented on Group-03/src/DataTracker.cpp:
```diff
@@ -5,104 +5,123 @@
-    // Create a copy of the vector and sort it
-    std::vector<double> sorted_values = values;
+    std::vector<T> sorted_values = values;
     std::sort(sorted_values.begin(), sorted_values.end());
```
A note alongside Amery's comment about nth_element, there are ranges implementations of both sort and nth_element that could be used here

Devon-FoxElster commented on Group-03/src/DataTracker.cpp:
```diff
@@ -5,104 +5,123 @@
+template <typename T>
+double DataTracker<T>::variance() const {
+    if (values.empty()) return 0.0;
+    if constexpr (std::is_arithmetic_v<T>) {
```
I like the use of the constexpr if

Devon-FoxElster commented on Group-03/src/DataTracker.cpp:
```diff
@@ -5,104 +5,123 @@
+
+// Determines if there is a winner (if any value surpasses 80% of total)
+template <typename T>
+std::optional<T> DataTracker<T>::winner() const {
+    if (values.empty()) return std::nullopt;
+
+    std::unordered_map<T, int> frequency;
+    size_t threshold = static_cast<size_t>(values.size() * 0.8);
+
+    for (const auto& val : values) {
+        if (++frequency[val] > static_cast<int>(threshold)) return val;
+
+    }
+
+    return std::nullopt;
```
I like the idea behind using std::optional here, but it doesn't look like this implementation is being explored at all on the testing side (in terms of checking whether it is a null optional value or not). Perhaps a test focused on stopping once a winner is found?

Devon-FoxElster commented:
# Use of Advanced C++ Features
### Classes Reviewed: DataTracker, Circle

Overall I saw constexpr implementation, use of lambdas, as well as type traits in the form of std::is_arithmetic_v. 
Constexpr was primarily used on if statements with the arithmetic check and looks good, as well as the lambda usages in Circle.

For the lambdas in Circle's Check proximity, I agree with Amery's comment about pulling the checkDistance one out as a member variable as there are applications for it in multiple functions. The second lambda for updating speed looks good.

The main recommendations I have are to look over all of your algorithms and decide whether you want to replace them with range counterparts, as they do look a bit cleaner to read. Unfortunately there is not a clear replacement for std::accumulate

Additionally, I would recommend using enums for Circle color, and possibly Circle characteristics if you have specific options in mind

Muelleralex1021 commented:
### Code Design

** Informative identifiers**

Muelleralex1021 commented:
### Code Design

**Informative identifiers**
Names of variables, names and types are clear and can be understood easily from reading them.

**Avoids “magic numbers”**
Saw a few magic numbers in the Circle class energy(100) eatingCounter >=5

**Intuitive and functional interface**
This is great!

**Informative comments**
Definitely lacking some comments that could be useful. 
DataTracker (I understand that the functions are doing what the name of them are. Maybe add calculating the mean of ...)
FixedPoint (comments are the same in header and implementation file. Consider adding more descriptive comments)
FunctionSet (comments are great!)
Sector (functions are pretty straightforward but maybe add some more description)
Surface(comments are great!)

**Consistent style**
Style is not consistent from class to class

**Visual appearance**
Visually the classes are easy to read through and look good.

Muelleralex1021 commented on Group-03/src/Circle.cpp:
```diff
@@ -2,10 +2,16 @@
 // Constructor
-Circle::Circle(double x, double y, double radius) : x_(x), y_(y), radius_(radius) {
+Circle::Circle(double x, double y, double radius, double baseSpeed, double speed, const std::string& circleType)
+    : x_(x), y_(y), radius_(radius), baseSpeed_(baseSpeed), speed_(speed), circleType_(circleType), energy_(100), 
```
What does the 100 here signify?

Muelleralex1021 commented on Group-03/src/Circle.cpp:
```diff
@@ -53,3 +91,71 @@ void Circle::setCharacteristic(const std::string& characteristic) {
+void Circle::eatPreyCircle() {
+    if (circleType_ == "red") {
+        eatingCounter_++;
+        if (eatingCounter_ >= 5) {
```
Magic Number?

Muelleralex1021 commented on Group-03/src/DataTracker.cpp:
```diff
@@ -5,104 +5,123 @@
-    
-    // Get the minimum element in the vector
-    double min = *std::min_element(values.begin(), values.end());
+// Calculates and returns the mode
```
Example here of descriptive comment. What are we calculating the mode of?

Muelleralex1021 commented:
Accidentally closed. So sorry!

Muelleralex1021 commented:
### Advanced C++ Classes

I agree with the Devon here on the use of constexpr, lambdas, and type traits like std::is_arithmetic_v—they're applied well for compile-time branching and help clarify intent. The lambdas in Circle::checkProximity are great for encapsulating logic. It does seem like extracting the distance calculation into a reusable member function or utility like stated before makes sense. The updateSpeed lambda is short and purpose-specific, so keeping it inline works well.

nigamsa1 commented on Group-03/src/DataTracker.cpp:
```diff
@@ -5,104 +5,123 @@
-    // Assertion to ensure that the minimum value is truly the smallest
-    assert(std::all_of(values.begin(), values.end(), [&](double v) { return v >= min; }) && 
-           "Min value is not actually the minimum.");
+    std::unordered_map<T, int> frequency;
```
I see your point but I do not think this would optimize the code enough to be worth it. Definitely something to consider for the future though and is a good suggestion!

vernonriley commented on Group-03/src/Circle.cpp:
```diff
@@ -53,3 +91,71 @@ void Circle::setCharacteristic(const std::string& characteristic) {
+    return repopulate_;
+}
+
+void Circle::eatPreyCircle() {
```
We are only using 2 types, red = pred and blue = prey. There is no future outlooks on adding more unless we work on it when the class is over.

mercere99 commented on Group-03/src/DataTracker.cpp:
```diff
@@ -5,104 +5,123 @@
-    // Create a copy of the vector and sort it
-    std::vector<double> sorted_values = values;
+    std::vector<T> sorted_values = values;
     std::sort(sorted_values.begin(), sorted_values.end());
```
Good catch!

mercere99 reviewed (approved):
Good job with `DataTracker` and `Circle`, though I have some remaining concerns.  The biggest is that it doesn't look like you responded to reviewers, so I'm not sure which comments you've fixed; some of them were serious.  For example, there seemed to be some concerns with DataTracker handling non-numeric types (though it looks like it should?).  Circle is more of a mess with some public member variables and duplicate variables declared (with and without an underscore).

I'm approving for now to merge in to the main branch (since it won't hurt any of the other projects), but I do recommend cleaning up the classes soon.

nstanis21 commented on Group-03/src/Sector.cpp:
```diff
@@ -5,19 +5,19 @@ namespace cse {
 }
 
-Circle* Sector::get_circle_at(size_t index) const {
+std::shared_ptr<Circle> Sector::get_circle_at(size_t index) const {
```
Could this be an std::optional to more explicitly show that sometimes a Circle may not be found at that index?

nstanis21 commented on Group-03/src/main.cpp:
```diff
@@ -0,0 +1,346 @@
+static constexpr int CHAR_SPACING = FONT_SCALE;
+
+// Digits
+static const uint8_t DIGIT[10][5] = {//chatgpt
```
Since the dimensions and content of this array is known, it should be able to be made into a constexpr as is, or using std::array

vernonriley replied:
Fixed that, thanks for pointing it out

nstanis21 commented on Group-03/src/main.cpp:
```diff
@@ -0,0 +1,346 @@
+};
+
+// Letters A, D, E, O, P, R, S, T, Y 
+static const uint8_t LETTER[26][5] = {//chatgpt
```
Same as above, could be made into a constexpr

vernonriley replied:
changed this as well.

nstanis21 commented on Group-03/src/main.cpp:
```diff
@@ -0,0 +1,346 @@
+    double cType1 = 0.0;
+    double cType2 = 0.0;
+
+    for (auto& circle : gCircles) {
```
You may be able to take advantage of std::ranges and its filtering function here

vernonriley replied:
We changed the way this works so it is no longer needed.

nstanis21 commented on Group-03/src/main.cpp:
```diff
@@ -0,0 +1,346 @@
+// Count circles of a given type
+int countCircles(const std::string& type)
+{
+    return std::count_if(gCircles.begin(), gCircles.end(),
```
Good use of lambdas here

vernonriley replied:
Thanks!

nstanis21 commented on Group-03/src/main.cpp:
```diff
@@ -0,0 +1,346 @@
+void drawString(const std::string& s, int x, int y, SDL_Color col)
+{
+    int cx = x;
+    for (size_t i = 0; i < s.size(); ++i) {
```
I think you could take advantage of std::views and its transformation features here

vernonriley replied:
Good idea, although I did struggle with trying to implement it. It does, however, seem slightly unnecessary so I didn't make any changes to it, just fixed an issue with changing it over to a constexpr.

nstanis21 commented:
Application Use of Modern C++ Review:
Good use of lambdas when applicable
Make sure to use constexpr when you can, I only found 2 cases, but you can probably find more since you're more familiar with the code
Check if there's places where you can take advantage of std::optional, I highlighted the case in Surfaces.cpp, but looking at the main.cpp code, your Circles class may also be able to take better advantage of it
Some places where you could have used std::ranges and std::view, but since most of your loops have predetermined lengths instead of being based on a containers size, I'm not sure how much benefit you can get from these outside the 2 cases I highlighted for you
For functions you are certain won't throw any errors, consider adding a 'noexcept' tag to them

nstanis21 commented on Group-03/src/main.cpp:
```diff
@@ -0,0 +1,346 @@
+        static Uint32 last = 0;
+        if (SDL_GetTicks() - last >= 1000) {
+            auto b = std::make_shared<Circle>(*vic);
+            gCircles.push_back(b);
```
Above you have to have a check for whether std::shared_ptr<Circle> is a nullptr or not. I feel like you could instead make a check here to make sure you don't add null pointers into gCircles since it should be overall safer compared to having gCircles hold a nullptr and implementing checks for it constantly

vernonriley replied:
This function actually makes the game not crash. We were running into issues of not having a timer on repopulation. The prey would repopulate infinitely and make the game crash. Nothing is being added to gCircles unless the 1 second timer has been completed from the last reproduction cycle.

nstanis21 commented:
Application Implementation Quality Review:
Lots of use of the magic strings "red" and "blue", I would recommend making these global variables
Add some extra checks to stop null pointers from being added to vectors/objects in the first place, instead of just checking once we access the object
You could put all your functions except for main() into a 'Board' class or similar, which would allow you or a user to more easily implement variations on the current application
Currently you're using binary values to define how each letter/number is constructed (the DIGIT and LETTER arrays), but this limits you to only 2 colors, I think you could get the same effect while just using normal integers, which would allow for more color variations on a board. This allows for more user customization, and can give some extra functionality for use cases or variations of the current application where more than one color might be needed
You are currently putting the arrays defining the characters in the DIGIT and LETTER arrays directly into the main.cpp file, instead it might be worth it to have them defined inside a .json file or something similar to allow for more flexibility

merryymary commented:
### Implementation

**No Warnings during compilation:**
I got these warnings when compiling with -Wall and -Wextra:

main.cpp:164:12: warning: variable 'cType1' set but not used [-Wunused-but-set-variable]
164 |     double cType1 = 0.0;
		
main.cpp:165:12: warning: variable 'cType2' set but not used [-Wunused-but-set-variable]
165 |     double cType2 = 0.0;

main.cpp:307:14: warning: unused parameter 'argc' [-Wunused-parameter]
307 | int main(int argc, char* argv[])

main.cpp:307:26: warning: unused parameter 'argv' [-Wunused-parameter]
307 | int main(int argc, char* argv[])


**Proper use of namespaces:**
	- Circle.cpp, Circle.h, FunctionSet.cpp, and main.cpp are not wrapped in namespace cse
	- All existing namespaces correctly do not include header files to avoid pollution
	- No using namespace std found in code, good job of not including that :D

**Adheres to the C++ standard**
	- Good job avoiding possible cases of undefined behavior, like dividing by 0. I specifically noticed it in DataTracker with checks to see if the values are not empty before doing calculations of dividing with values.size()

**Appropriate Modularity**
	- In Surface.cpp, there are a lot of static casting of sector coordinates. Although it’s one line, you could put that into a helper function.
	- In Surface.cpp in update() and move_circle(), there seems to be the same if check and remove_circle to add_circle logic. If that is the case, you could make a helper function of doing the remove/add cycle.
	- Other than that, the code is very well factored and delegated to functions to reduce repeated code

**Uses modern c++ constructs**
	- One technique we talked about in class was the spaceship operator <=> which could be used to create some of the default operators in FixedPoint. 
	- Good use of constexpr on the scale() function in FixedPoint
	- There are indexed for loops in Surface.cpp and not ranged, but I can see the reasoning behind this due to needing the indexes for some calculations.
	- In Sector.cpp, get_circle_at returns a nullptr, this could be a use for std::optional. DataTracker does a good job of using this.
	- Surface.cpp has a good use of lambdas in check_collision()

**Uses standard libray algorithms**
	- DataTracker does good job of using std::accumulate, std::sort, and std::min/max element
	- In DataTracker delete_value, this has some of essentially the same functionality as remove_if, although I don’t know how much complexity would be reduced in that function. Just something to think about I guess!
	- Good use of std::sort in DataTracker
	- Good use of std::clamp in Surface.cpp

**Pointer Handling:** All pointer instances are shared_ptr!
**Const Correctness:** All const functions are correctly marked as well as paramaters.

### TESTING AND ERROR CHECKING

**Circle**
	- When running Circle.cpp code I got this error:    26 |     double angle = (std::rand() / static_cast<double>(RAND_MAX)) * 2 * M_PI;
	- CircleAdvanced won’t compile for me because of creating circles with “red” parameter instead of CircleType
	- I think you should combine all test cases for circle into one class
	- In testing, setCircleType, getCircleType, eatPreyCircle are not the correct names for your current functions.
	- GetType(), isRegenerating(), hasSpeedBoost(), setRadius(), setType(), regenerateEnergy() aren't checked
	- Good edge cases with energy and radius checking
	Circle returns errors for trying to access private member variables and constructor errors
	
**DataTracker**
	- DataTracker.cpp doesn't check variance(), winner() or has(). DataTrackerAdvanced checks variance() and winner(), but not has(). I would combine these into one cpp file.
	- Good testing off various types that DataTracker supports
	- If you wanted to cover everything, you could duplicated the DataTrackerAdvanced classes for other types like string for example

**Fixed Point**
	- Add cases that include negative values, probably to all of the test cases.
	- Good job having test cases for all public functions and operators

**Function Set**
	- Can add a case for removing a function that doesn't exist
	- Could check what happens when you call CallAll on an empty function set

**Surface**
	- Update() and check_collision() not tested
	- Could add a case for remove circle that doesn't exist in the surface or the sector
	- Could implement a test for if you add the same exact circle to the surface

Overall, this project was really dope and I thought it ran really smoothly in the actual application on my local. The code was really well documented and made it easy to understand and read through, so as a code reviewer thank ya'll. I literally tried to be as nitpicky as possible with suggestions since you guys did a good job overall.

merryymary commented on Group-03/src/FunctionSet.cpp:
```diff
@@ -112,18 +116,20 @@ int cse::FunctionSet<R, Args...>::FindFunctionIndex(const FunctionType& func) co
-    }
+        if (func == nullptr)
+{
+    throw std::invalid_argument("AddFunction: Function target is null.");
```
Should say RemoveFunction :D

merryymary commented on Group-03/src/Surface.cpp:
```diff
@@ -1,107 +1,146 @@
+      surface_size(sector_size),
+      sector_width(width  / sector_size),
+      sector_height(height / sector_size)
+{
```
You could add a check here to ensure surface_size is > 0 before creating the object. I just noticed surface_size is used a lot in division calculation and could lead to UB of dividing by zero potentially

nigamsa1 replied:
Just made it so it would throw an error if it is 0 or smaller

merryymary commented on tests/Group-03/Circle.cpp:
```diff
@@ -4,44 +4,46 @@
+    CHECK(mag > 0);
+
+    Circle blue(0,0,3,2,2,"blue");
+    CHECK(blue.canRepopulate());
```
Just how you check if blue can repopulate, I would also recommend checking that red correctly returns that it doesn't.  Essentially in these constructors it's not a bad idea to check that every variable that's set in the constructor is set and returned correctly.

nigamsa1 replied:
Just added this in. Good suggestion!

merryymary commented on tests/Group-03/Circle.cpp:
```diff
@@ -4,44 +4,46 @@
+    auto circle = std::make_shared<Circle>(15.0, 25.0, 5.0, 2, 2, "red");
+
+    surface.add_circle(circle);
+    surface.move_circle(circle, 35.0, 45.0);
```
I would add some test cases with negative values here :D

nigamsa1 replied:
Added!

merryymary commented on tests/Group-03/CircleAdvanced.cpp:
```diff
@@ -1,69 +1,209 @@
+
+TEST_CASE("Circle basic construction and mutators", "[Circle]") {
+
+    Circle c(10.0, 15.0, kRadius, kBaseSpeed, kInitSpeed, "red");
```
I don't know if it's just my compiler being funky, but it wouldn't compile these Circle classes. I believe it's because in the construct you use CircleType of Predator or Prey and here you use "red" or "blue" as the indicator.

nigamsa1 replied:
Fixed!

merryymary commented on tests/Group-03/DataTracker.cpp:
```diff
@@ -1,88 +1,109 @@
+    CHECK(tracker.min() == 0);
+    CHECK(tracker.max() == 0);
+    CHECK(tracker.mode() == 0);
 }
```
If you have a tracker of a type double, and then try to add something of a different type, like string, what happens?

nigamsa1 replied:
adding different types leads to a compiler error something along the lines of "no matching function for call to cse::DataTracker<double>::add_value(std::string)"

merryymary commented on tests/Group-03/DataTrackerAdvanced.cpp:
```diff
@@ -28,18 +39,16 @@ TEST_CASE("Advanced DataTracker Variance", "[DataTracker]") {
     CHECK(tracker.variance() == Approx(4.0));
```
I would add cases with negatives as well as additional tests of variance on empty, variance with 1 input.

nigamsa1 replied:
Added these in

merryymary commented on tests/Group-03/FixedPoint.cpp:
```diff
@@ -4,136 +4,136 @@
 
 // Constructor Tests
 TEST_CASE("FixedPoint Default Constructor", "[FixedPoint]") {
-    Fixed1 fp;
```
I would go through and add some negative test cases on these just to ensure it works properly on positive and negative.

nigamsa1 replied:
added a whole other set of tests that are the same as the ones that exist, but for negative numbers

merryymary commented on tests/Group-03/Surface.cpp:
```diff
@@ -4,37 +4,38 @@
+        cse::Surface surface(100, 100, 10);
+        auto circle = std::make_shared<Circle>(0, 0, 1, 1, 1, "red");
+        surface.add_circle(circle);
+        surface.move_circle(circle, 15, 15);
```
Could check move on a circle that is created but hasn't been added

nigamsa1 replied:
Added a test case for this!

nigamsa1 commented:
@merryymary 
I removed everything that was unused that you saw in the warnings!

I also wrapped every file in namespace cse. Thank you for catching that.

A helper function for static casting would definitely be a good idea! If we have time at the end I will for sure implement that so it is easier to read.

added std::optional to get_circle_at


TESTING AND ERROR CHECKING

I fixed this, M_PI was not defined: When running Circle.cpp code I got this error: 26 | double angle = (std::rand() / static_cast(RAND_MAX)) * 2 * M_PI;

I completely redid the Circle tests so it tests everything and compiles correctly all in one file. We changed the code but not the tests! Sorry about that!

Fixed DataTracker tests to be in one file and added tests for has()

added negative test cases to FixedPoint

ze-ntsh commented on tests/Group-03/Surface.cpp:
```diff
@@ -1,40 +1,57 @@
+using cse::Circle;
+using cse::CircleType;
+using std::make_shared;
+
```
Add tests specifically for collision detection:
It should test collision of moving circles especially in boundary conditions. Example claude generated snipped.
```c++
TEST_CASE("SurfaceTest 4: Collision Detection", "[Surface]") {
    Surface surface(100, 100, 10);
    auto circle1 = make_shared<Circle>(10.0, 10.0, 5.0, 1.0, CircleType::Predator);
    auto circle2 = make_shared<Circle>(19.0, 10.0, 5.0, 1.0, CircleType::Prey);
    
    surface.add_circle(circle1);
    surface.add_circle(circle2);
    
    auto collision = surface.check_collision();
    CHECK(collision.first == "add");
    CHECK((collision.second.first == circle1 && collision.second.second == circle2) || 
          (collision.second.first == circle2 && collision.second.second == circle1));
}
```

ze-ntsh commented on tests/Group-03/Surface.cpp:
```diff
@@ -1,40 +1,57 @@
+    REQUIRE_NOTHROW(surface.move_circle(circle, 20.0, 30.0));
+    CHECK(circle->getX() == Approx(20.0));
+    CHECK(circle->getY() == Approx(30.0));
+}
```
These seem to be few tests for your main class. It would be good seeing some more tests in here for it to be reliable especially pertaining to predator prey interactions and the energy system

ze-ntsh reviewed (commented):
**Testing & Error Checking**  
I've already added inline comments on the main test files you presented in class — overall, they look good. A few suggestions to further strengthen the test suite:
- **Test edge cases more thoroughly**, especially in `DataTracker` (e.g., division by zero errors).
- **Expand collision detection tests** by introducing more moving particles and adding more scenarios for predator-prey interactions within `Surface`.
- **Consolidate redundant test sections** to make the test suite more organized and maintainable.

These improvements would make your testing more robust and better validate the expected behavior of your classes.

---

**User Experience**  
The UI is simple and currently offers a fairly linear user experience — which is not necessarily a bad thing. A few ideas you might consider:
- Allow users to **select the starting number of circles** for each category. (I understand you mentioned this is heavy on computation at the moment, but it could be a good future enhancement.)
- **Make the winning team text more prominent** — maybe larger, and consider adding the total time the simulation ran.
- Introduce **basic user controls** like pausing or adjusting parameters mid-simulation to increase engagement.

Again, I don’t expect these to be added immediately, but they’re worth considering for the future if you continue building on this project. Overall, it's entertaining to watch the simulation run!

ze-ntsh commented on Group-03/src/Circle.cpp:
```diff
@@ -1,55 +1,115 @@
+    if (y_ - radius_ < 0 || y_ + radius_ > height) dy_ = -dy_;
+}
+
+void Circle::decreaseEnergy(int amount) {
```
I am not sure if decreased energy can be negative. So, add in a check for that. However, if that is expected behavior, ignore this. 
```c++
if (amount < 0) {
    throw std::invalid_argument("Cannot decrease energy by negative amount");
}
```

ze-ntsh commented on Group-03/src/Surface.cpp:
```diff
@@ -1,107 +1,152 @@
+    int new_y = static_cast<int>(y_pos) / surface_size;
+
+    if (old_x != new_x || old_y != new_y) {
+        sectors[old_x][old_y].remove_circle(circle);
```
Here this can be out of bounds. You can use .at() or maybe check bounds beforehand. I know these values are obtained internally but what if some other part mutated the sectors.

ze-ntsh commented on Group-03/src/DataTracker.cpp:
```diff
@@ -1,108 +1,152 @@
+    return std::nullopt;
+}
+
+template class cse::DataTracker<int>;
```
While these template instantiations do show which types are expected but there's no validation that others won't be used (if this class was to be reused somewhere else)

ze-ntsh commented on Group-03/src/Sector.cpp:
```diff
@@ -1,26 +1,26 @@
+}
+
+void Sector::remove_circle(std::shared_ptr<Circle> circle) {
+    circles.erase(std::remove(circles.begin(), circles.end(), circle), circles.end());
```
A nullptr check could be good here

gr005100 reviewed (commented):
Code Design

Informative identifiers: Variables are very clear and are consistent through out your code

Avoids "magic numbers": Everything looks good I don't see any use of magic numbers

Intuitive and functional interface: Id suggest in FunctionSet.h to avoid doing args... and &&... callArgs this is awfully confusing for a anyone who isn't aware of everything and or is new to c++ as well as naming auto CallAll(Args... args) isn't very definitive 

Informative comments: There are plenty of comments however Id suggest adding some comments in Circle.cpp to offer more detail on whats been done. A suggestion would be to add more comments on the large member initializer list and how ChatGPT was used.

Consistent style: Great work you follow PascalCase through out everything

Visual appearance: Visually everything is well structured  in white space, indention, and line breaks especially in the member initializer list. 


Use of Modern C++

Regarding your use of modern C++ great job you use multiple templates, stl algorithms such as begin, end, make pair, empty, and many more more. A couple suggestions would be to potentially get rid of the nested for loops and do it in a much more optimal.  


Overall great work

Thisisnotgoingtowork commented:
> Quote reply
Thank you for your comments!
We fix most of the things you said about the main, yet for the Surface, we check the point that you mention, and we did conclude that at this moment it is not that necessary to do so, still, thank you for pointing that out!

Thisisnotgoingtowork commented:
> Application Implementation Quality Review: Lots of use of the magic strings "red" and "blue", I would recommend making these global variables
> 
> Add some extra checks to stop null pointers from being added to vectors/objects in the first place, instead of just checking once we access the object
> 
> You could put all your functions except for main() into a 'Board' class or similar, which would allow you or a user to more easily implement variations on the current application
> 
> Currently you're using binary values to define how each letter/number is constructed (the DIGIT and LETTER arrays), but this limits you to only 2 colors, I think you could get the same effect while just using normal integers, which would allow for more color variations on a board. This allows for more user customization, and can give some extra functionality for use cases or variations of the current application where more than one color might be needed
> 
> You are currently putting the arrays defining the characters in the DIGIT and LETTER arrays directly into the main.cpp file, instead it might be worth it to have them defined inside a .json file or something similar to allow for more flexibility

They are all fixed. But for the .json part, at this point, since we are using the emcc to form the page, the .json file is generated automatically, in which I don't suggest to change that, just in case it won't break the whole app. But thanks for pointing that out, if next time we are not using emcc and having the .json file, we can do that of course.

Thisisnotgoingtowork commented:
> **Testing & Error Checking** I've already added inline comments on the main test files you presented in class — overall, they look good. A few suggestions to further strengthen the test suite:
> 
> * **Test edge cases more thoroughly**, especially in `DataTracker` (e.g., division by zero errors).
> * **Expand collision detection tests** by introducing more moving particles and adding more scenarios for predator-prey interactions within `Surface`.
> * **Consolidate redundant test sections** to make the test suite more organized and maintainable.
> 
> These improvements would make your testing more robust and better validate the expected behavior of your classes.
> 
> **User Experience** The UI is simple and currently offers a fairly linear user experience — which is not necessarily a bad thing. A few ideas you might consider:
> 
> * Allow users to **select the starting number of circles** for each category. (I understand you mentioned this is heavy on computation at the moment, but it could be a good future enhancement.)
> * **Make the winning team text more prominent** — maybe larger, and consider adding the total time the simulation ran.
> * Introduce **basic user controls** like pausing or adjusting parameters mid-simulation to increase engagement.
> 
> Again, I don’t expect these to be added immediately, but they’re worth considering for the future if you continue building on this project. Overall, it's entertaining to watch the simulation run!

I fixed Most of part that you comment, and there are some I don't think matters, or at this point is needed. But overall, thank you for the comments.
PS: I did reply to each of different comments you put, but I don't know why it showing Pending and not shown to others.

hdseyfferth reviewed (commented):
### Final Review
**User Experience**
Accesibility: 
- There were a long list of instructions to be able to open and run the application. In the scope of a cse course that’s fine, but if you wanted to continue developing further, I would suggest making it more accesible.
- My group used github’s hosting feature and it was pretty easy to step up and I believe it is also free: https://pages.github.com/

Ease of use / Intuitiveness:
Comments:
- The app is straight forward and easy to use. It opens up the the predator vs pray simulation and runs automatically so there is no set up work to be done once it is running.
- The display of the simulation is front and center so the main application future is easy to find and intuitively placed.
- The animation looks great and has very little lag and updates seamlessly to the next iteration of the simulation.

Suggestions:
I have a few simple / small suggestions for making the UI more intuitive:
- Have a start / stop feature for the simulation. (Could be simple buttons)
- Show stats at beginning or during the simulation. This would allow users to see a summary of the details behind the simulation. It would also show off some of the work your code is doing behind the seens. Could include some live stats as well.
- Add a small popup or text display that breifly summarizes what the program is simulating and how it works.
- Add a small dialog that states what round it is. This could potentially lead to a displayed list so you can see information about how many rounds have occured and who won each.

Bugs:
- The menu of options at the top (fullscreen, resize, lock/hide mouse) do not appear to work on my end. I am using chrome on a macbook.

**Code Design**
Overal structure is well thought out and readable. The style of the code and naming conventions were consistent across classes. There are some small improvements that can be made that I commented on inline. Here is a short summary of small improvements:
- avoid magic numbers
- add more informative comments (for ex: in DataTracker when using algorithms / lambdas) to explain the purpose behind them.
- use encapsulation where applicable to make files more concise + easier to follow

Since I am getting to these reviews a bit late in the game, I will put this review under 'Comment' instead of 'Request Changes' to hopefully prevent you from having to do more changes at this point. Great project - good job everyone :)

garciaow commented:
> Code Design
> 
> Informative identifiers: Variables are very clear and are consistent through out your code
> 
> Avoids "magic numbers": Everything looks good I don't see any use of magic numbers
> 
> Intuitive and functional interface: Id suggest in FunctionSet.h to avoid doing args... and &&... callArgs this is awfully confusing for a anyone who isn't aware of everything and or is new to c++ as well as naming auto CallAll(Args... args) isn't very definitive
> 
> Informative comments: There are plenty of comments however Id suggest adding some comments in Circle.cpp to offer more detail on whats been done. A suggestion would be to add more comments on the large member initializer list and how ChatGPT was used.
> 
> Consistent style: Great work you follow PascalCase through out everything
> 
> Visual appearance: Visually everything is well structured in white space, indention, and line breaks especially in the member initializer list.
> 
> Use of Modern C++
> 
> Regarding your use of modern C++ great job you use multiple templates, stl algorithms such as begin, end, make pair, empty, and many more more. A couple suggestions would be to potentially get rid of the nested for loops and do it in a much more optimal.
> 
> Overall great work

Thank you for your review! You're absolutely right about FunctionSet; I'll see to these suggestions for the base classes and will look into optimizing the draw and update functions further.

nigamsa1 commented on Group-03/src/index.html:
This file is actually created by emscripten and is not something we coded!

nigamsa1 commented:
> ### Final Review
> **User Experience** Accesibility:
> 
> * There were a long list of instructions to be able to open and run the application. In the scope of a cse course that’s fine, but if you wanted to continue developing further, I would suggest making it more accesible.
> * My group used github’s hosting feature and it was pretty easy to step up and I believe it is also free: https://pages.github.com/
> 
> Ease of use / Intuitiveness: Comments:
> 
> * The app is straight forward and easy to use. It opens up the the predator vs pray simulation and runs automatically so there is no set up work to be done once it is running.
> * The display of the simulation is front and center so the main application future is easy to find and intuitively placed.
> * The animation looks great and has very little lag and updates seamlessly to the next iteration of the simulation.
> 
> Suggestions: I have a few simple / small suggestions for making the UI more intuitive:
> 
> * Have a start / stop feature for the simulation. (Could be simple buttons)
> * Show stats at beginning or during the simulation. This would allow users to see a summary of the details behind the simulation. It would also show off some of the work your code is doing behind the seens. Could include some live stats as well.
> * Add a small popup or text display that breifly summarizes what the program is simulating and how it works.
> * Add a small dialog that states what round it is. This could potentially lead to a displayed list so you can see information about how many rounds have occured and who won each.
> 
> Bugs:
> 
> * The menu of options at the top (fullscreen, resize, lock/hide mouse) do not appear to work on my end. I am using chrome on a macbook.
> 
> **Code Design** Overal structure is well thought out and readable. The style of the code and naming conventions were consistent across classes. There are some small improvements that can be made that I commented on inline. Here is a short summary of small improvements:
> 
> * avoid magic numbers
> * add more informative comments (for ex: in DataTracker when using algorithms / lambdas) to explain the purpose behind them.
> * use encapsulation where applicable to make files more concise + easier to follow
> 
> Since I am getting to these reviews a bit late in the game, I will put this review under 'Comment' instead of 'Request Changes' to hopefully prevent you from having to do more changes at this point. Great project - good job everyone :)

Thank you for the review Hannah! I definitely agree that I do not think our app is the easiest to get running. Ideally, in the future, we would be able to deploy the application using a more intuitive hosting platform. 

I also think those are great features that I will try to implement if I work on this further!

The menu of options at the top (fullscreen, resize, lock/hide mouse) are feautures of emscripten and not our app itself. I was also never able to get those working haha but they are just emscripten options and we did not create those.

I will look for magic numbers and I will look into those comments if I have time today but I may not be able to get to them! I will follow up more if I am able to.
